Services/Simulations.cs

```diff
@@ -192,7 +192,19 @@ public async Task<Models.Simulation> UpsertAsync(Models.Simulation simulation)
                 throw new InvalidInputException("Simulation ID is not specified.");
             }
 
-            var existingSimulation = await this.GetAsync(simulation.Id);
+            Models.Simulation existingSimulation = null;
```

Comment:
We shouldn't get in to a state where UpsertAsync is being called for a simulation that doesn't exist. InsertAsync should be called. Under what scenario are we seeing that UpsertAsync is being called for a simulation that doesn't exist? I could see that possibly we'd get in to a race condition where a simulation was deleted right before being upserted, but in that case we should not re-create the simulation.

Reply:
resolved offline